### PR TITLE
WSTEAMA-347: Migrate ATIAnalytics component

### PR DESCRIPTION
**Pull request by karinathomasbbc**
Resolves https://jira.dev.bbc.co.uk/browse/WSTEAMA-347

Overall changes
======
Migrates the ATIAnalytics component to src/app/components and upgrades to TypeScript.

Code changes
======
- Move component to src/app/components & update all import references
- Convert all files to typescript

Testing
======
- [x] PR checks passing

Helpful Links
======
[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: src/app/components/ATIAnalytics/params/index.ts [hunk 1]
```diff
@@ -95,48 +121,91 @@ const pageTypeParamBuilders = {
...
+    requestContext: RequestContextProps,
+    serviceContext: ServiceConfig,
+    contentType?: string,
+  ): ATIPageTrackingProps;
+};
+
+const createBuilderFactory = (
+  requestContext: RequestContextProps,
+  pageTypeHandlers: {
+    [key: string]: UrlBuilderFunction | ParamBuilderFunction;
```
Comment (karinathomasbbc) [inline src/app/components/ATIAnalytics/params/index.ts]: Should we try t type the key to `PageTypes` somehow?

File: src/app/components/ATIAnalytics/beacon/index.test.ts [hunk 2]
```diff
@@ -1,35 +1,30 @@
-import * as sendBeacon from '#lib/analyticsUtils/sendBeacon';
-import * as analyticsUtils from '#lib/analyticsUtils';
+import * as sendBeacon from '../../../lib/analyticsUtils/sendBeacon';
+import * as analyticsUtils from '../../../lib/analyticsUtils';
 import { sendEventBeacon } from '.';
 
 const sendBeaconSpy = jest.spyOn(sendBeacon, 'default');
+
+// @ts-expect-error - we need to mock these functions to ensure tests are deterministic
 analyticsUtils.getAtUserId = jest.fn().mockReturnValue('123-456-789');
```
Comment (amoore108) [inline src/app/components/ATIAnalytics/beacon/index.test.ts]: Just for these expect-errors, you may be able to wrap the mocked functions in something like:
`(analyticsUtils.getAtUserId as jest.Mock) = ...` to cast it to a Mock type.

File: src/app/components/ATIAnalytics/params/cpsAssetPage/buildParams.ts [hunk 3]
```diff
@@ -4,24 +4,27 @@ import {
...
   const { atiAnalyticsAppName, atiAnalyticsProducerId, service, brandName } =
     serviceContext;
 
   const { metadata, promo } = pageData;
 
-  const getChapter1 = pageIdentifier => {
+  const getChapter1 = (pageIdentifier: string) => {
     if (service === 'news') {
-      return path(['atiAnalytics', 'chapter'], metadata);
+      return metadata?.atiAnalytics?.chapter;
```
Comment (amoore108) [inline src/app/components/ATIAnalytics/params/cpsAssetPage/buildParams.ts:27]: Nice, like seeing the reduction in Ramda usage.

**Review by amoore108 — APPROVED** [review]
Looks great! Thanks for doing this 🙏 Just a small comment on the test expect-error use.

File: src/app/components/ATIAnalytics/params/cpsAssetPage/buildParams.ts (hunk 3 above)
Comment (karinathomasbbc) [reply src/app/components/ATIAnalytics/params/cpsAssetPage/buildParams.ts:27]: Yup, removing the ramda `path` functionality made the typing a lot easier - I didn't need to cast each variable `as string`

File: src/app/components/ATIAnalytics/beacon/index.test.ts (hunk 2 above)
Comment (karinathomasbbc) [reply src/app/components/ATIAnalytics/beacon/index.test.ts]: Oh brill, thanks for this - will do that now.

**Comment by karinathomasbbc** [comment]
Converting to draft & marking as do not merge until we have the outcome of the spike to create a new PianoAnalytics component

**Comment by pvaliani** [comment]
Outcome of spike:

- New PianoAnalytics component not being created
- Spike was successful in showing that metadata for ATI can be returned via BFF
- PR to return this data will be merged, this PR will also be merged and then Simorgh HomePage will be modified in a subsequent PR (branched from here: https://github.com/bbc/simorgh/pull/10819) to return the HomePage ATI data in Simorgh via the BFF.

File: src/app/components/ATIAnalytics/index.test.tsx [hunk 4]
```diff
@@ -68,6 +73,7 @@ describe('ATI Analytics Container', () => {
   describe('pageType article', () => {
     it('should call CanonicalATIAnalytics when platform is canonical', () => {
       const mockCanonical = jest.fn().mockReturnValue('canonical-return-value');
+      // @ts-expect-error - we need to mock these functions to ensure tests are deterministic
```
Comment (andrewscfc) [inline src/app/components/ATIAnalytics/index.test.tsx:76]: Can you explain this, I don't understand why this is necessary?

File: src/app/components/ATIAnalytics/index.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+const ATIAnalytics = ({ data }: ATIProps) => {
+  const requestContext = useContext(RequestContext);
+  const serviceContext = useContext(ServiceContext);
+  const { isAmp } = requestContext;
+
+  const pageviewParams = buildATIUrl(
+    data,
+    requestContext,
+    serviceContext,
+  ) as string;
```
Comment (andrewscfc) [inline src/app/components/ATIAnalytics/index.tsx:18]: Is the `as` necessary here, isn't the type just inferred from the `buildATIUrl` function?

File: src/app/components/ATIAnalytics/params/index.ts [hunk 6]
```diff
@@ -145,10 +225,12 @@ export const buildATIEventTrackingParams = (
     );
 
     return buildParams(data, requestContext, serviceContext);
-  } catch (error) {
+  } catch (error: unknown) {
```
Comment (andrewscfc) [inline src/app/components/ATIAnalytics/params/index.ts:228]:
```suggestion
  } catch (error: Error) {
```
Any reason we can't type this here? I assume there is but just interested what's going on here

**Review by andrewscfc — APPROVED** [review]
Looks good, a few questions for my understanding I think 👍

File: src/app/components/ATIAnalytics/index.test.tsx (hunk 4 above)
Comment (pvaliani) [reply src/app/components/ATIAnalytics/index.test.tsx:76]: My understanding is that our general usage of the `@ts-expect-error` directive here and in other areas of this particular work is temporary and will be removed as part of the work to clean up the component. The directive here isn't required in order to allow the test to pass. I can remove it if you like, however there may be a reason it's been added that i'm unaware of which will inform upcoming maintenance or tech work as it's also used elsewhere in this test and others.

I think that the way this test is mocking the `Canonical` platform attempts to modify the `default` which is read-only and the directive is suppressing that error. Perhaps the intention here is to maintain the way it's currently mocked to then be refactored later? Apologies if that sounds a bit wooly!

File: src/app/components/ATIAnalytics/index.tsx (hunk 5 above)
Comment (pvaliani) [reply src/app/components/ATIAnalytics/index.tsx:18]: `pageviewParams` is defined as a string here. However, removing the `as string` as suggested then creates a type error on this prop being passed to the `amp` and `canonical` components. Making this change at the function level  and making `buildATIUrl` a string type causes a similar issue so I feel like the type isn't easily as inferred as I was expecting it to be. [https://github.com/bbc/simorgh/blob/bfbbfe1a9e89ce7583a01c28e2acd2c8c6966e2f/src/app/components/ATIAnalytics/types.ts#L65](URL)

**Pull request merged**